rust/src/sdk/api.rs

```diff
@@ -0,0 +1,542 @@
+}
+
+#[allow(unused_unsafe)]
+#[no_mangle]
+/// Creates a new [`Client`].
+///
+/// Takes a `period` in seconds for which the [`Client`] will try to poll the coordinator for new
+/// broadcasted FL round data.
+///
+/// Takes a `callback` function pointer and a void pointer to the `state` of the callback. The
+/// underlying function is defined over void pointers referencing (anonymous) structs for the
+/// `state` and `input` arguments of the callback.
+///
+/// # Safety
+/// The method depends on the safety of the `callback` and on the consistent definition and layout
+/// of its `input` across the FFI-boundary.
+pub unsafe extern "C" fn new_client(
+    period: c_ulong,
+    callback: unsafe extern "C" fn(*mut c_void, *const c_void),
+    state: *mut c_void,
+) -> *mut Client {
+    if period == 0 {
+        // TODO: add error handling
+        panic!("polling period must be positive")
+    }
+    let client = if let Ok(client) = Client::new(period as u64) {
+        client
+    } else {
+        // TODO: add error handling
+        panic!("participant initialization failed")
+    };
+
+    // TODO: actually start the client, requires tokio running
+
+    #[repr(C)]
+    struct Input {
+        _round_started: bool,
+        _participant_initialized: bool,
+        _model_cached: bool,
+    }
+    let input = &Input {
+        _round_started: client.has_new_coord_pk_since_last_check,
+        _participant_initialized: true,
+        _model_cached: client.cached_model.is_some(),
+    } as *const _ as *const c_void;
+    unsafe {
+        // safe if the `callback` is sound and the same definition and layout is used for `Input`
+        // across the FFI-boundary by the caller
+        callback(state, input)
+    };
+
+    Box::into_raw(Box::new(client))
+}
+
+#[allow(unused_unsafe)]
+#[no_mangle]
+/// Destroys a [`Client`] and frees its allocated memory.
+///
+/// # Safety
+/// The method dereferences from the raw pointer arguments. Therefore, the behavior of
+/// the method is undefined if the arguments don't point to valid objects.
+pub unsafe extern "C" fn drop_client(client: *mut Client) {
+    if !client.is_null() {
+        unsafe {
+            // safe if the raw pointer `client` comes from a valid allocation of a `Client`
+            Box::from_raw(client);
+        }
+    }
+}
+
+#[allow(unused_unsafe)]
+#[no_mangle]
+/// Checks if the next round has started.
+///
+/// # Safety
+/// The method dereferences from the raw pointer arguments. Therefore, the behavior of
+/// the method is undefined if the arguments don't point to valid objects.
+pub unsafe extern "C" fn is_next_round(client: *mut Client) -> bool {
+    if client.is_null() {
+        // TODO: add error handling
+        panic!("invalid client");
+    }
+    let client = unsafe {
+        // safe if the raw pointer `client` comes from a valid allocation of a `Client`
+        &mut *client
+    };
+    let is_next_round = client.has_new_coord_pk_since_last_check;
```

Comment:
Just an idea: We could also use Option<()> here and write:
```
let is_next_round = client.has_new_coord_pk_since_last_check.take();
```

Reply:
Or another solution 
```
let is_next_round = std::mem::replace(client.has_new_coord_pk_since_last_check, false);
```

Reply:
good ideas, i'll change that.